Added support for FCM V1 configuration

edi opened the pull request:
Based on this FCM related blogpost here, [https://expo.dev/blog/expo-adds-support-for-fcm-http-v1-api](url) it seems like there's no solution for people like us who use the Node SDK to send push notifications.

This change adds support for initiating `ExpoClient` with another optional parameter called `useFCM: true` which appends the necessary query string to the `/push/send` (?useFcmV1=true)

p19ky commented:
This is needed!

edi commented:
@p19ky 

> This is needed!

Unsure if anyone actually checks this repo anymore, but you can always use `patch-package` to fix (until they merge this PR)

akravecas8 commented:
Hi can we use this now?

edi commented:
> Hi can we use this now?

Unfortunately you cannot, until a maintainer merges it.

akravecas8 commented:
@edi 

I've removed my Legacy Server Key from the Expo Credentials console; shouldn't it automatically default to using FCMv1 service account key anyway? It's not, but shouldn't it?

ide commented on README.md:
```diff
@@ -16,7 +16,10 @@ import { Expo } from 'expo-server-sdk';
-let expo = new Expo({ accessToken: process.env.EXPO_ACCESS_TOKEN });
+let expo = new Expo({
+  accessToken: process.env.EXPO_ACCESS_TOKEN,
+  useFCM: false // this can be set to true in order to use the FCM v1 API
```
```suggestion
  useFCMv1: false // this can be set to true in order to use the FCM v1 API
```

edi replied:
You're succesting we should use useFCMv1 as prop? Because that also means I should change it everywhere else, and not just here.

ide commented on src/ExpoClient.ts:
```diff
@@ -75,13 +77,13 @@ export class Expo {
    * sized chunks.
    */
   async sendPushNotificationsAsync(messages: ExpoPushMessage[]): Promise<ExpoPushTicket[]> {
+    const useFCM = this.useFCM ? '?useFcmV1=true' : '';
```
Could you use the `URL` API instead? With even more string concatenation, the risk of strings being treated as URL components goes up.

Also, it would be good to have the logic be: if the option is set to true or false, set the query parameter, otherwise omit the query parameter and let the server choose its default behavior. That way people can explicitly opt out of using FCM v1 too.

edi replied:
Makes sense, will sort tomorrow.

edi replied:
Just realised our usernames are an anagram 😄

edi reviewed (commented):
Finished using URL API.
Also renamed the `useFcmV1` config var to match the QS lettercasing.

edi commented on src/ExpoClient.ts:
```diff
@@ -75,13 +77,14 @@ export class Expo {
    */
   async sendPushNotificationsAsync(messages: ExpoPushMessage[]): Promise<ExpoPushTicket[]> {
+    const url = new URL(`${BASE_API_URL}/push/send`);
+    if (this.useFcmV1) url.searchParams.append('useFcmV1', 'true');
```
Left these inline to make the prettier config happy.

edi commented:
@ide by the way, something is wrong with the push notification API.
The `default behavior` which is supposed to use legacy api (if legacy credential exists, otherwise use fcm v1) doesn't seem to work.

For example, if I remove my legacy key (while having set up an fcm v1 key), the API returns an `InvalidCredentials` error, stating that:

`Unable to retrieve the FCM server key for the recipient's app. Make sure you have provided a server key as directed by the Expo FCM documentation.`

Just as [akravecas8](https://github.com/akravecas8) mentioned as well, I was unsure but tested now.

If however I enable "force fcm v1", then the notification is sent as usual.

Conclusion, the default behavior doesn't seem to "fallback" to FCMv1 if a legacy api key is not found.

edi commented:
Couldn't edit the above post, but here you can see a screenshot of my current configuration.

1. Added FCM v1 key
2. Removed legacy
3. Used the **web push tool** to test, and it comes up with an error on both "default behavior" as well as "force legacy". However, "force fcm v1" works just fine.


Screenshot of configuration: [here](https://github.com/expo/expo-server-sdk-node/assets/7712869/6f4e18b8-7b6b-4378-bd5c-2bcc670b9297)
Screenshot of error: [here](https://github.com/expo/expo-server-sdk-node/assets/7712869/618e197a-2797-4a05-94c8-3791d67a99fd)

ide commented:
@christopherwalter could you look into the above?

christopherwalter commented:
@edi thanks for flagging this issue with the default behavior! I've confirmed that there's a bug in production, and I have a fix ready that will land before end of today.

jikez commented:
Hi @christopherwalter I'm wondering the issue for this has been fixed regarding `The default behavior which is supposed to use legacy api (if legacy credential exists, otherwise use fcm v1) doesn't seem to work.` ? 
I'm testing the new key right now, but without passing in `{ useFcmV1: true }`, it's not using the v1 key
